### PR TITLE
[noderesourcetopology] move logging to the logr interface

### LINKED ISSUE #709: [NodeResourceTopology] review and overhaul internal logging
The logging inside the NodeResourceTopology code grew over time without a real design and lacking a clear architecture.
The plan is to fix this. I'm proposing to review all the usages of `klog` over the codebase (roughly 110 at my last count and)

1. make sure either the `namespace/name` pair or the `UID' of the pod being processed are logged, to make as easy as possibly to track the flows while reviewing the logs (either by humans or by machines, either online, offline, manually or using log ingestion tools)
2. extending the above, make the logs consistents as possible. Always use structured logging (this means getting rid of the few instances of `Warningf` in the codebase)
3. enable log pluggability. The code should use a `logr.Logger` instance, not `klog` directly. Doing this way, we lose nothing because `klog` integrates natively with `logr` (see: https://pkg.go.dev/k8s.io/klog/v2#pkg-overview) and of course by default the code will still log through `klog` (using https://pkg.go.dev/k8s.io/klog/v2#Background). But we will enable trivial log replacement should be needed by integrators, for example for better auditing of the logging decisions.

Note: point 3 will _not_ require any extra dep, since both `scheduler-plugins` and `k/k` already depend on `go-logr`, and AFAICT there are no plans on the horizon to remove this dependency.
Note: point 3 will also be an enabler for point 1 above (see: https://pkg.go.dev/k8s.io/klog/v2#LoggerWithValues)

### PR DESCRIPTION
**Pull request by ffromani**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Implement the changes proposed in #709 to review logging and make it consistent, easy to extend and to get right, and for bonus points easy to replace should it be needed.

Worth pointing out this PR goes in the direction of KEP 3077 (contextual logging). The code becomes almost completely ready for a switch to the contextual logger, with only few selected pieces to change in the future.

#### Which issue(s) this PR fixes:
Fixes #709 

#### Special notes for your reviewer:
PR split in many smallish commit to make it easier to review (and bisect)
Excluding the obvious intended logging fixes/improvements, no intended changes in behavior.
Ultimately, this PR is not user-facing because the users should see no difference except for more consistent logs, easier to grep and/or process in general.

#### Does this PR introduce a user-facing change?
```release-note
NONE
```

### CODE REVIEW

**Comment by ffromani** [comment]
/hold

The PR is reviewable but I'm still polishing it a bit and validating the changes (slicing the logs with grep mostly)

File: pkg/noderesourcetopology/cache/overreserve.go [hunk 1]
```diff
@@ -54,23 +56,24 @@ type OverReserve struct {
...
+	resyncMethod := getCacheResyncMethod(lh, cfg)
 
 	nrtObjs := &topologyv1alpha2.NodeResourceTopologyList{}
 	// TODO: we should pass-in a context in the future
 	if err := client.List(context.Background(), nrtObjs); err != nil {
 		return nil, err
 	}
 
-	klog.V(3).InfoS("nrtcache: initializing", "objects", len(nrtObjs.Items), "method", resyncMethod)
+	lh.V(3).Info("initializing", "objects", len(nrtObjs.Items), "method", resyncMethod)
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/cache/overreserve.go:72]: maybe change objects to something more descriptive like `nrt count` or just `nrts`

File: pkg/noderesourcetopology/cache/overreserve.go [hunk 2]
```diff
@@ -97,45 +100,45 @@ func (ov *OverReserve) GetCachedNRTCopy(ctx context.Context, nodeName string, po
 		return nrt, true
 	}
 
-	klog.V(6).InfoS("nrtcache NRT", "logID", klog.KObj(pod), "vanilla", stringify.NodeResourceTopologyResources(nrt))
+	ov.lh.V(6).Info("NRT", "logID", klog.KObj(pod), "vanilla", stringify.NodeResourceTopologyResources(nrt))
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/cache/overreserve.go:103]: s/vanilla/actual ?

File: pkg/noderesourcetopology/cache/passthrough.go [hunk 3]
```diff
@@ -19,30 +19,32 @@ package cache
...
 	}
 }
 
 func (pt Passthrough) GetCachedNRTCopy(ctx context.Context, nodeName string, _ *corev1.Pod) (*topologyv1alpha2.NodeResourceTopology, bool) {
-	klog.V(5).InfoS("Lister for nodeResTopoPlugin")
+	pt.lh.V(5).Info("Lister for nodeResTopoPlugin")
 	nrt := &topologyv1alpha2.NodeResourceTopology{}
 	if err := pt.client.Get(ctx, types.NamespacedName{Name: nodeName}, nrt); err != nil {
-		klog.V(5).ErrorS(err, "Cannot get NodeTopologies from NodeResourceTopologyLister")
+		pt.lh.V(5).Error(err, "Cannot get NodeTopologies from NodeResourceTopologyLister")
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/cache/passthrough.go:47]: lowercase?

File: pkg/noderesourcetopology/config.go [hunk 4]
```diff
@@ -81,17 +81,17 @@ func updateTopologyManagerConfigFromAttributes(conf *TopologyManagerConfig, attr
...
 
 func updateTopologyManagerConfigFromTopologyPolicies(conf *TopologyManagerConfig, nodeName string, topologyPolicies []string) {
+	lh := logging.Log()
 	if len(topologyPolicies) == 0 {
-		klog.V(3).InfoS("Cannot determine policy", "node", nodeName)
 		return
 	}
 	if len(topologyPolicies) > 1 {
-		klog.V(4).InfoS("Ignoring extra policies", "node", nodeName, "policies count", len(topologyPolicies)-1)
+		lh.V(4).Info("Ignoring extra policies", "node", nodeName, "policies count", len(topologyPolicies)-1)
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/config.go:89]: lowercase?

File: pkg/noderesourcetopology/filter.go [hunk 5]
```diff
@@ -41,74 +42,71 @@ const highestNUMAID = 8
 
 type PolicyHandler func(pod *v1.Pod, zoneMap topologyv1alpha2.ZoneList) *framework.Status
 
-func singleNUMAContainerLevelHandler(pod *v1.Pod, zones topologyv1alpha2.ZoneList, nodeInfo *framework.NodeInfo) *framework.Status {
-	klog.V(5).InfoS("Single NUMA node handler")
+func singleNUMAContainerLevelHandler(lh logr.Logger, pod *v1.Pod, zones topologyv1alpha2.ZoneList, nodeInfo *framework.NodeInfo) *framework.Status {
+	lh.V(5).Info("Single NUMA node handler")
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/filter.go:46]: lowercase?

File: pkg/noderesourcetopology/filter.go [hunk 6]
```diff
@@ -41,74 +42,71 @@ const highestNUMAID = 8
...
-	logNumaNodes("container handler NUMA resources", nodeInfo.Node().Name, nodes)
+	logNumaNodes(lh, "container handler NUMA resources", nodeInfo.Node().Name, nodes)
 
 	// the init containers are running SERIALLY and BEFORE the normal containers.
 	// https://kubernetes.io/docs/concepts/workloads/pods/init-containers/#understanding-init-containers
 	// therefore, we don't need to accumulate their resources together
 	for _, initContainer := range pod.Spec.InitContainers {
-		logID := fmt.Sprintf("%s/%s/%s", pod.Namespace, pod.Name, initContainer.Name)
-		klog.V(6).InfoS("target resources", stringify.ResourceListToLoggable(logID, initContainer.Resources.Requests)...)
+		lh.V(6).Info("target resources", stringify.ResourceListToLoggable(initContainer.Resources.Requests)...)
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/filter.go:59]: maybe reword `target resources` to `init container resources` ?

File: pkg/noderesourcetopology/filter.go [hunk 7]
```diff
@@ -41,74 +42,71 @@ const highestNUMAID = 8
...
+			lh.V(2).Info("cannot align container", "name", initContainer.Name, "kind", "init")
 			return framework.NewStatus(framework.Unschedulable, "cannot align init container")
 		}
 	}
 
 	for _, container := range pod.Spec.Containers {
-		logID := fmt.Sprintf("%s/%s/%s", pod.Namespace, pod.Name, container.Name)
-		klog.V(6).InfoS("target resources", stringify.ResourceListToLoggable(logID, container.Resources.Requests)...)
+		// TODO: add containerName
+		lh.V(6).Info("target resources", stringify.ResourceListToLoggable(container.Resources.Requests)...)
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/filter.go:71]: s/target resources/container resources ?

File: pkg/noderesourcetopology/filter.go [hunk 8]
```diff
@@ -175,20 +173,19 @@ func isResourceSetSuitable(qos v1.PodQOSClass, resource v1.ResourceName, quantit
 	return numaQuantity.Cmp(quantity) >= 0
 }
 
-func singleNUMAPodLevelHandler(pod *v1.Pod, zones topologyv1alpha2.ZoneList, nodeInfo *framework.NodeInfo) *framework.Status {
-	klog.V(5).InfoS("Pod Level Resource handler")
+func singleNUMAPodLevelHandler(lh logr.Logger, pod *v1.Pod, zones topologyv1alpha2.ZoneList, nodeInfo *framework.NodeInfo) *framework.Status {
+	lh.V(5).Info("Pod Level Resource handler")
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/filter.go:177]: lowercase?

File: pkg/noderesourcetopology/filter.go [hunk 9]
```diff
@@ -204,30 +201,33 @@ func (tm *TopologyMatch) Filter(ctx context.Context, cycleState *framework.Cycle
...
-		klog.V(2).InfoS("invalid topology data", "node", nodeName)
+		lh.V(2).Info("invalid topology data")
 		return framework.NewStatus(framework.Unschedulable, "invalid node topology data")
 	}
 	if nodeTopology == nil {
 		return nil
 	}
 
-	klog.V(5).InfoS("Found NodeResourceTopology", "nodeTopology", klog.KObj(nodeTopology))
+	lh.V(5).Info("Found NodeResourceTopology", "nodeTopology", klog.KObj(nodeTopology))
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/filter.go:216]: lowercase?

File: pkg/noderesourcetopology/cache/store.go [hunk 10]
```diff
@@ -142,7 +146,7 @@ func (rs *resourceStore) UpdateNRT(logID string, nrt *topologyv1alpha2.NodeResou
 				if zr.Available.Cmp(qty) < 0 {
 					// this should happen rarely, and it is likely caused by
 					// a bug elsewhere.
-					klog.V(3).InfoS("nrtcache: cannot decrement resource", "logID", logID, "zone", zr.Name, "node", nrt.Name, "available", zr.Available, "requestor", key, "quantity", qty)
+					rs.lh.V(3).Info("cannot decrement resource", "logID", logID, "zone", zr.Name, "node", nrt.Name, "available", zr.Available, "requestor", key, "quantity", qty)
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/cache/store.go:149]: s/qty/qty.String() ?

File: pkg/noderesourcetopology/plugin.go [hunk 11]
```diff
@@ -127,9 +133,9 @@ func New(args runtime.Object, handle framework.Handle) (framework.Plugin, error)
 		return nil, err
 	}
 
-	nrtCache, err := initNodeTopologyInformer(tcfg, handle)
+	nrtCache, err := initNodeTopologyInformer(lh, tcfg, handle)
 	if err != nil {
-		klog.ErrorS(err, "Cannot create clientset for NodeTopologyResource", "kubeConfig", handle.KubeConfig())
+		lh.Error(err, "Cannot create clientset for NodeTopologyResource", "kubeConfig", handle.KubeConfig())
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/plugin.go:138]: lowercase?

File: pkg/noderesourcetopology/podprovider/podprovider.go [hunk 12]
```diff
@@ -43,38 +43,38 @@ func NewFromHandle(handle framework.Handle, cacheConf *apiconfig.NodeResourceTop
...
 }
 
-func IsPodRelevantDedicated(pod *corev1.Pod, logID string) bool {
+func IsPodRelevantDedicated(lh logr.Logger, pod *corev1.Pod, logID string) bool {
 	// Every other phase we're interested into (see https://github.com/kubernetes-sigs/scheduler-plugins/pull/599).
 	// Note PodUnknown is deprecated and reportedly no longer set since 2015 (!!)
 	if pod.Status.Phase == corev1.PodPending {
 		// this is unexpected, so we're loud about it
-		klog.V(2).InfoS("nrtcache: Listed pod in Pending phase, ignored", "logID", logID, "podUID", pod.UID)
+		lh.V(2).Info("nrtcache: Listed pod in Pending phase, ignored", "logID", logID, "podUID", pod.UID)
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/podprovider/podprovider.go:72]: remove `nrtcache:` ?

File: pkg/noderesourcetopology/podprovider/podprovider.go [hunk 13]
```diff
@@ -43,38 +43,38 @@ func NewFromHandle(handle framework.Handle, cacheConf *apiconfig.NodeResourceTop
...
 		return false
 	}
 	if pod.Spec.NodeName == "" {
 		// this is very unexpected, so we're louder about it
-		klog.InfoS("nrtcache: Listed pod unbound, ignored", "logID", logID, "podUID", pod.UID)
+		lh.Info("nrtcache: Listed pod unbound, ignored", "logID", logID, "podUID", pod.UID)
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/podprovider/podprovider.go:77]: remove `nrtcache:` ?

File: pkg/noderesourcetopology/podprovider/podprovider.go [hunk 14]
```diff
@@ -43,38 +43,38 @@ func NewFromHandle(handle framework.Handle, cacheConf *apiconfig.NodeResourceTop
 	podInformer := coreinformers.NewFilteredPodInformer(handle.ClientSet(), metav1.NamespaceAll, 0, cache.Indexers{}, nil)
 	podLister := podlisterv1.NewPodLister(podInformer.GetIndexer())
 
-	klog.V(5).InfoS("Start custom pod informer")
+	lh.V(5).Info("Start custom pod informer")
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/podprovider/podprovider.go:46]: lowercase?

File: pkg/noderesourcetopology/podprovider/podprovider.go [hunk 15]
```diff
@@ -43,38 +43,38 @@ func NewFromHandle(handle framework.Handle, cacheConf *apiconfig.NodeResourceTop
...
 	ctx := context.Background()
 	go podInformer.Run(ctx.Done())
 
-	klog.V(5).InfoS("Syncing custom pod informer")
+	lh.V(5).Info("Syncing custom pod informer")
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/podprovider/podprovider.go:50]: lowercase?

File: pkg/noderesourcetopology/podprovider/podprovider.go [hunk 16]
```diff
@@ -43,38 +43,38 @@ func NewFromHandle(handle framework.Handle, cacheConf *apiconfig.NodeResourceTop
...
 	cache.WaitForCacheSync(ctx.Done(), podInformer.HasSynced)
-	klog.V(5).InfoS("Synced custom pod informer")
+	lh.V(5).Info("Synced custom pod informer")
```
Comment (PiotrProkop) [inline pkg/noderesourcetopology/podprovider/podprovider.go:52]: lowercase?

File: pkg/noderesourcetopology/podprovider/podprovider.go (hunk 13 above)
Comment (ffromani) [reply pkg/noderesourcetopology/podprovider/podprovider.go:77]: done, thanks.

File: pkg/noderesourcetopology/podprovider/podprovider.go (hunk 12 above)
Comment (ffromani) [reply pkg/noderesourcetopology/podprovider/podprovider.go:72]: done, thanks.

File: pkg/noderesourcetopology/podprovider/podprovider.go (hunk 15 above)
Comment (ffromani) [reply pkg/noderesourcetopology/podprovider/podprovider.go:50]: fixed, thanks

File: pkg/noderesourcetopology/podprovider/podprovider.go (hunk 16 above)
Comment (ffromani) [reply pkg/noderesourcetopology/podprovider/podprovider.go:52]: fixed, thanks

File: pkg/noderesourcetopology/podprovider/podprovider.go (hunk 14 above)
Comment (ffromani) [reply pkg/noderesourcetopology/podprovider/podprovider.go:46]: fixed, thanks

File: pkg/noderesourcetopology/cache/overreserve.go (hunk 2 above)
Comment (ffromani) [reply pkg/noderesourcetopology/cache/overreserve.go:103]: will go for "unmodified"

File: pkg/noderesourcetopology/cache/passthrough.go (hunk 3 above)
Comment (ffromani) [reply pkg/noderesourcetopology/cache/passthrough.go:47]: thanks, fixed

File: pkg/noderesourcetopology/config.go (hunk 4 above)
Comment (ffromani) [reply pkg/noderesourcetopology/config.go:89]: thanks, fixed

File: pkg/noderesourcetopology/filter.go (hunk 5 above)
Comment (ffromani) [reply pkg/noderesourcetopology/filter.go:46]: thanks, fixed

File: pkg/noderesourcetopology/filter.go (hunk 8 above)
Comment (ffromani) [reply pkg/noderesourcetopology/filter.go:177]: thanks, fixed

File: pkg/noderesourcetopology/filter.go (hunk 9 above)
Comment (ffromani) [reply pkg/noderesourcetopology/filter.go:216]: thanks, fixed

File: pkg/noderesourcetopology/plugin.go (hunk 11 above)
Comment (ffromani) [reply pkg/noderesourcetopology/plugin.go:138]: thanks, fixed

File: pkg/noderesourcetopology/cache/overreserve.go (hunk 1 above)
Comment (ffromani) [reply pkg/noderesourcetopology/cache/overreserve.go:72]: using `noderesourcetopologies`

File: pkg/noderesourcetopology/cache/overreserve.go (hunk 2 above)
Comment (ffromani) [reply pkg/noderesourcetopology/cache/overreserve.go:103]: decided to change both: "vanilla" -> "fromcache"; "updated" -> "withassumed"

File: pkg/noderesourcetopology/filter.go (hunk 6 above)
Comment (ffromani) [reply pkg/noderesourcetopology/filter.go:59]: thanks, rephrased

File: pkg/noderesourcetopology/filter.go (hunk 7 above)
Comment (ffromani) [reply pkg/noderesourcetopology/filter.go:71]: thanks, rephrased

File: pkg/noderesourcetopology/cache/store.go (hunk 10 above)
Comment (ffromani) [reply pkg/noderesourcetopology/cache/store.go:149]: this seems better, thanks

**Comment by ffromani** [comment]
the test failures are known. It seems better to me to push some code down the stack: https://github.com/k8stopologyawareschedwg/noderesourcetopology-api/pull/35

**Comment by ffromani** [comment]
/hold cancel

Bar fixes and review comments, I think this PR has all the content we need for this cleanup.

**Comment by PiotrProkop** [comment]
Should we squash history to 2 commits?
- bump nrt
- rework logging?

**Comment by ffromani** [comment]
> Should we squash history to 2 commits?
> 
>     * bump nrt-api
> 
>     * rework logging?

Consider how the project seems to prefer, I'm in favor of squashing once reviewers are happy.

**Comment by ffromani** [comment]
in the past couple of uploads I fixed nits and inconsistencies here and there. The PR is reviewable and "done" (I can send more fixups in followup PRs).
Actuallly I also have a couple of extensions already to propose building on this work.

If reviewers feel a squashed PR is more manageable I'll be upload in no time.

**Comment by PiotrProkop** [comment]
LGTM! Great work! After squash I can properly approve the PR 🚀

File: pkg/noderesourcetopology/least_numa.go [hunk 17]
```diff
@@ -37,6 +37,7 @@ const (
 )
 
 func leastNUMAContainerScopeScore(pod *v1.Pod, zones topologyv1alpha2.ZoneList) (int64, *framework.Status) {
+	lh := logging.Log()
```
Comment (Tal-or) [inline pkg/noderesourcetopology/least_numa.go]: Is there a specific reason why we should call the logger at the begging of each function? or its just used as a shortcut?

Comment (ffromani) [reply pkg/noderesourcetopology/least_numa.go]: it's a transitional change. In this phase I'm doing the minimal change to move from global logger to `logr.Logger` instances. Most of these usages will be removed in the upcoming changes.

File: pkg/noderesourcetopology/cache/discardreserved.go [hunk 18]
```diff
@@ -46,12 +46,14 @@ type DiscardReserved struct {
...
 	client         ctrlclient.Client
+	lh             logr.Logger
 }
 
-func NewDiscardReserved(client ctrlclient.Client) Interface {
+func NewDiscardReserved(lh logr.Logger, client ctrlclient.Client) Interface {
 	return &DiscardReserved{
 		client:         client,
 		reservationMap: make(map[string]map[types.UID]bool),
+		lh:             lh,
```
Comment (Tal-or) [inline pkg/noderesourcetopology/cache/discardreserved.go:57]: I noticed the logger is being used often. maybe it's worth embedding it on all the structs to save some key strokes?

Comment (ffromani) [reply pkg/noderesourcetopology/cache/discardreserved.go:57]: I have a strong bias towards composition rather than embedding, but if all the reviewers feel differently I'm open to reconsider

**Comment by ffromani** [comment]
/hold

need to review the integration with the scheduler framework >= 1.29

**Comment by ffromani** [comment]
/hold cancel

proper integration with framework and ongoing work to fully support the contextual logging (KEP-3077) would need a followup PR.
Let's stop with this already huge PR right here, I feel this is a massive improvement (I know I'm biased here :) ) already.

changelog:
1. added constants for well-known keys for structured logging
2. annotated all the plugin entry points like
```
	lh.V(4).Info(logging.FlowBegin)
	defer lh.V(4).Info(logging.FlowEnd)
```

**Comment by ffromani** [comment]
now I call this PR done :)

**Comment by PiotrProkop** [comment]
Looks good to me 👍 
/lgtm
